File: backend/app/models/batter_ability.rb
class BatterAbility < ApplicationRecord
    RANKS = { G: 0, F: 1, E: 2, D: 3, C: 4, B: 5, A: 6, S: 7 }.freeze      

    # バリデーション
    validates :trajectory, presence: true, inclusion: 0..4
    validates :hit, presence: true, inclusion: 0..100
    validates :power, presence: true, inclusion: 0..100
    validates :run_speed, presence: true, inclusion: 0..100
    validates :arm_strength, presence: true, inclusion: 0..100
    validates :fielding, presence: true, inclusion: 0..100
    validates :catching, presence: true, inclusion: 0..100
    validates :clutch, inclusion: { in: ['G','F','E','D','C','B','A','S'], allow_blank: true }
    validates :vs_lhp, inclusion: { in: ['G','F','E','D','C','B','A','S'], allow_blank: true }
    validates :stearing, inclusion: { in: ['G','F','E','D','C','B','A','S'], allow_blank: true}
    validates :runnning, inclusion: { in: ['G','F','E','D','C','B','A','S'], allow_blank: true}
    validates :throwing, inclusion: { in: ['G','F','E','D','C','B','A','S'], allow_blank: true }
    validates :catcher, inclusion: { in: ['G','F','E','D','C','B','A','S'], allow_blank: true}
    validates :grit, inclusion: { in: ['G','F','E','D','C','B','A','S'], allow_blank: true}
    validates :recovery, inclusion: { in: ['G','F','E','D','C','B','A','S'], allow_blank: true }
    validates :clutch_rank, :vs_lhp_rank, :stealing_rank, :running_rank, :throwing_rank, :grit_rank, :recovery_rank, presence: true

    enum clutch_rank:   RANKS, _prefix: true
    enum vs_lhp_rank:   RANKS, _prefix: true
    enum stealing_rank: RANKS, _prefix: true
    enum running_rank:  RANKS, _prefix: true
    enum throwing_rank: RANKS, _prefix: true
    enum catcher_rank:  RANKS, _prefix: true
    enum grit_rank:     RANKS, _prefix: true
    enum recovery_rank: RANKS, _prefix: true

    # アソシエーション
    belongs_to :player_season
end

File: frontend/src/models/batter_ability.ts
// ドメインモデル
export type BatterAbility = {
  trajectory: number | null;
  hit: number | null;
  power: number | null;
  run_speed: number | null;
  arm_strength: number | null;
  fielding: number | null;
  catching: number | null;

  clutch_rank: string | null;
  vs_lhp_rank: string | null;
  stealing_rank: string | null;
  running_rank: string | null;
  throwing_rank: string | null;
  catcher_rank: string | null;
  grit_rank: string | null;
  recovery_rank: string | null;
  special_ability?: string | null;
};

export const createBatterAbility = (overrides?: Partial<BatterAbility>): BatterAbility => ({
  trajectory: null,
  hit: null,
  power: null,
  run_speed: null,
  arm_strength: null,
  fielding: null,
  catching: null,
  clutch_rank: 'D',
  vs_lhp_rank: 'D',
  stealing_rank: 'D',
  running_rank: 'D',
  throwing_rank: 'D',
  catcher_rank: 'D',
  grit_rank: 'D',
  recovery_rank: 'D',
  special_ability: '',
  ...overrides,
});

// 送信フォーム用
export type BatterAbilityForm = {
  trajectory: string;
  hit: string;
  power: string;
  run_speed: string;
  arm_strength: string;
  fielding: string;
  catching: string;
  clutch_rank: string;
  vs_lhp_rank: string;
  stealing_rank: string;
  running_rank: string;
  throwing_rank: string;
  catcher_rank: string;
  grit_rank: string;
  recovery_rank: string;
  special_ability: string;
};

export const buildBatterAbilityPayload = (
  form: BatterAbilityForm
): BatterAbility => ({
  trajectory: form.trajectory ? Number(form.trajectory) : 0,
  hit: form.hit ? Number(form.hit) : 0,
  power: form.power ? Number(form.power) : 0,
  run_speed: form.run_speed ? Number(form.run_speed) : 0,
  arm_strength: form.arm_strength ? Number(form.arm_strength) : 0,
  fielding: form.fielding ? Number(form.fielding) : 0,
  catching: form.catching ? Number(form.catching) : 0,
  clutch_rank: form.clutch_rank || 'D',
  vs_lhp_rank: form.vs_lhp_rank || 'D',
  stealing_rank: form.stealing_rank || 'D',
  running_rank: form.running_rank || '',
  throwing_rank: form.throwing_rank || 'D',
  catcher_rank: form.catcher_rank || 'D',
  grit_rank: form.grit_rank || 'D',
  recovery_rank: form.recovery_rank || 'D',
  special_ability: form.special_ability || null,
});